processor/lsmintervalprocessor/internal/merger/value.go

```diff
@@ -0,0 +1,796 @@
+		overflowSum.SetAggregationTemporality(pmetric.AggregationTemporalityDelta)
+		overflowDP := overflowSum.DataPoints().AppendEmpty()
+		if err := decorate(
+			overflowDP.Attributes(),
+			s.scopeDPLimitCfg.Overflow.Attributes,
+		); err != nil {
+			return pmetric.Metrics{}, fmt.Errorf("failed to finalize merged metric: %w", err)
+		}
+		overflowDP.SetIntValue(int64(sm.datapointsLimits.EstimateOverflow()))
+	}
+	// Remove any hanging resource or scope which failed to have any entries
+	// due to children reaching their limits.
+	// TODO (lahsivjar): We can probably optimize to not require this loop by
+	// adding to source metric only at finalize.
+	s.source.ResourceMetrics().RemoveIf(func(rm pmetric.ResourceMetrics) bool {
+		rm.ScopeMetrics().RemoveIf(func(sm pmetric.ScopeMetrics) bool {
+			sm.Metrics().RemoveIf(func(m pmetric.Metric) bool {
+				switch m.Type() {
+				case pmetric.MetricTypeGauge:
+					return m.Gauge().DataPoints().Len() == 0
+				case pmetric.MetricTypeSum:
+					return m.Sum().DataPoints().Len() == 0
+				case pmetric.MetricTypeHistogram:
+					return m.Histogram().DataPoints().Len() == 0
+				case pmetric.MetricTypeExponentialHistogram:
+					return m.ExponentialHistogram().DataPoints().Len() == 0
+				case pmetric.MetricTypeSummary:
+					return m.Summary().DataPoints().Len() == 0
+				}
+				return false
+			})
+			return sm.Metrics().Len() == 0
+		})
+		return rm.ScopeMetrics().Len() == 0
+	})
+	return s.source, nil
+}
+
+// merge merges the provided value to the current value instance.
+func (v *Value) merge(op Value) error {
+	for mOtherID, mOther := range op.metricLookup {
+		resOtherID := mOtherID.Resource()
+		scopeOtherID := mOtherID.Scope()
+		resOther := op.resLookup[resOtherID]
+		scopeOther := op.scopeLookup[scopeOtherID]
+
+		// Merge/add resource metrics. Note that if the resource metrics
+		// overflows then the ID will be different from the other resource ID.
+		resID, err := v.addResourceMetrics(resOther.ResourceMetrics)
+		if err != nil {
+			return fmt.Errorf("failed to merge resource metrics: %w", err)
+		}
+		// Merge/add scope metrics. Note that if the scope metrics overflows
+		// then the ID will be different from the other scope ID.
+		scopeID, err := v.addScopeMetrics(resID, scopeOther.ScopeMetrics)
+		if err != nil {
+			return fmt.Errorf("failed to merge scope metrics: %w", err)
+		}
+		// Merge any overflow estimators for scope or datapoints. Note that here
+		// we assume that the limits for both metrics being merged are identical
+		// and thus if any of the metric has overflowed then the target metric
+		// for merge will definitely overflow. Thus, merging the estimators is
+		// safe and required to correctly estimate the total number of overflow.
+		// Note that overflow merging is not deterministic and the estimates are
+		// always estimates as it is possible that the oveflows have hashes which
+		// haven't overflowed in the source metric.
+		res := v.resLookup[resID]
```

Comment:
Should addResourceMetrics return resourceMetrics rather than just the identity? Then we don't need to do a separate lookup. Same for addScopeMetrics & the scopeLookup below

Reply:
Done! Also, did a bit of similar changes around the add datapoint methods.